src/PatternBuilder.ts

```diff
@@ -0,0 +1,381 @@
+import * as fc from 'fast-check'
+import { pipe } from 'fp-ts/function'
+import * as RA from 'fp-ts/ReadonlyArray'
+import { match, matchOn } from './internal/match'
+
+// Breakdown of RegExp from https://262.ecma-international.org/5.1/#sec-15.10
+// Slightly adapted
+
+/** @internal */
+export type Pattern = Disjunction | TermSequence | Term
+
+type Disjunction = { tag: 'disjunction'; left: Pattern; right: TermSequence | Term }
+
+type TermSequence = { tag: 'termSequence'; terms: ReadonlyArray<Term> }
+
+type Term = Atom | QuantifiedAtom
+
+type QuantifiedAtom = { tag: 'quantifiedAtom'; atom: Atom; greedy: boolean } & (
+  | { kind: 'star' }
+  | { kind: 'plus' }
+  | { kind: 'question' }
+  | { kind: 'exactly'; count: number }
+  | { kind: 'minimum'; min: number }
+  | { kind: 'between'; min: number; max: number }
+)
+
+type Atom = { tag: 'atom' } & (
+  | { kind: 'character'; char: string }
+  | { kind: 'anything' }
+  | {
+      kind: 'characterClass'
+      exclude: boolean
+      ranges: ReadonlyArray<{ lower: number; upper: number }>
+    }
+  | { kind: 'subgroup'; subpattern: Pattern }
+)
+
+type Char = string
+
+const matchK = matchOn('kind')
+
+/** @since 1.0.0 */
+export const char = (c: Char): Atom => ({ tag: 'atom', kind: 'character', char: c })
+
+/** @since 1.0.0 */
+export const anything: Atom = { tag: 'atom', kind: 'anything' }
+
+/** @since 1.0.0 */
+export const characterClass = (
+  exclude: boolean,
+  ...ranges: ReadonlyArray<readonly [Char, Char]>
+): Atom => ({
+  tag: 'atom',
+  kind: 'characterClass',
+  exclude,
+  ranges: ranges.map(
+    ([c1, c2]) => ({ lower: c1.charCodeAt(0), upper: c2.charCodeAt(0) } as const)
+  ),
+})
```

Comment:
:wrench: this needs some tweaking; it should accept single characters as well as numeric ranges